### PR TITLE
Ingress configuration options

### PR DESCRIPTION
**Pull request by hydrogen18**
This gives users the options in their SDL to configure

1. read timeout
2. send timeout
3. body size limits
4. number of retries (for when replicas > 1)
5. retry cases
6. retry timeout

These become annotations that control our kubernetes NGINX ingress controller. There are some limits on these values, for example I have it so body size can't be more than 50 MB right now.

If this isn't specified in the SDL the defaults are the same as what is in effect today on everyone's ingress controller.

### CODE REVIEW

**Review by dmikey — APPROVED** [review]
I would be interested to see perf testing against the ingress controller with different body loads. My overall question is how large and how many simultaneous bodies can be uploaded before degraded perf across the controller. I'm assuming this controller is one per cluster, and a provider is responsible for that cluster configuration and performance. How do we measure that for quality of service?

**Comment by hydrogen18** [comment]
The NGINX controller is an instance set at present, so it is indeed one per cluster. We can move to a daemonset in the future if the need arises to get more performance.

At present I think our actual bottleneck is the internal networking of the cluster, we probably should spend more time to make sure we're getting the most out of the infrastructure we have with our networking setup.

File: sdl/v2.go [hunk 1]
```diff
@@ -24,11 +52,92 @@ type v2ExposeTo struct {
...
+		nextCases = defaultNextCases
+	} else {
+		for _, nextCase := range nextCases {
+			switch nextCase {
+			case nextCaseOff:
+				if len(nextCases) != 1 {
+					return manifest.ServiceExposeHTTPOptions{}, errCannotSpecifyOffAndOtherCases
+				}
+			case nextCaseError:
+			case nextCaseTimeout:
```
Comment (troian) [inline sdl/v2.go:119]: missing `fallthrough` ?

**Review by troian — CHANGES_REQUESTED** [review]
some tests are failing as well

```
--- FAIL: TestV2HTTPOptionsParse (0.00s)
    v2_test.go:337: 
        	Error Trace:	v2_test.go:337
        	Error:      	Received unexpected error:
        	            	open _testdata/simple_httpoptions.yaml: no such file or directory
        	Test:       	TestV2HTTPOptionsParse
```

File: sdl/v2.go (hunk 1 above)
Comment (hydrogen18) [reply sdl/v2.go:119]: It should take no action in the event that this case is hit, which is the correct thing

**Review by 0x0I — APPROVED** [review]

File: provider/cluster/kube/builder.go [hunk 2]
```diff
@@ -944,11 +944,56 @@ func ingressHost(lid mtypes.LeaseID, svc *manifest.Service) string {
...
+	// For kubernetes/ingress-nginx
+	// https://github.com/kubernetes/ingress-nginx
+	const root = "nginx.ingress.kubernetes.io"
+
+	httpOptions := b.expose.HTTPOptions
+
+	result := map[string]string{
+
+		fmt.Sprintf("%s/proxy-send-timeout", root):        fmt.Sprintf("%dms", httpOptions.SendTimeout),
+		fmt.Sprintf("%s/proxy-read-timeout", root):        fmt.Sprintf("%dms", httpOptions.ReadTimeout),
```
Comment (boz) [inline provider/cluster/kube/builder.go:957]: should we just have them input it as a string `"100ms"`?

I wonder what will happen if we have invalid values here.

File: sdl/v2.go [hunk 3]
```diff
@@ -24,11 +52,92 @@ type v2ExposeTo struct {
...
+		ho.ReadTimeout != 0 ||
+		ho.SendTimeout != 0 ||
+		ho.NextTries != 0 ||
+		ho.NextTimeout != 0
+}
+
+func (ho v2HTTPOptions) asManifest() (manifest.ServiceExposeHTTPOptions, error) {
+	maxBodySize := ho.MaxBodySize
+	if 0 == maxBodySize {
+		maxBodySize = defaultMaxBodySize
```
Comment (boz) [inline sdl/v2.go:84]: these defaults would ideally be configurable

**Review by boz — COMMENTED** [review]
Looks great!

Will this be compatible with current, saved manifests?

Should we put a limit on some of these values?

**Review by boz — COMMENTED** [review]
I see the limits now.

This is super cool

File: sdl/v2.go (hunk 3 above)
Comment (hydrogen18) [reply sdl/v2.go:84]: On chain or per provider ?

File: provider/cluster/kube/builder.go (hunk 2 above)
Comment (hydrogen18) [reply provider/cluster/kube/builder.go:957]: The input type is `uint32` so they can't really input anything invalid. If they input zero we use the default. A positive number is a valid value, up to the limit.

**Comment by hydrogen18** [comment]
> Looks great!
> 
> Will this be compatible with current, saved manifests?
> 
> Should we put a limit on some of these values?

It is compatible with saved manifest because it is additive only. Omitting this section results in values that are the same as the defaults today

File: sdl/v2.go (hunk 3 above)
Comment (boz) [reply sdl/v2.go:84]: I think per provider.

forget it, this is fine.

**Comment by troian** [comment]
lgtm. one more little think. there are non-gofmt file(s) https://github.com/ovrclk/akash/pull/1323/checks?check_run_id=3085774438

**Comment by jay-dee7** [comment]
@hydrogen18 @boz @dmikey guys this PR looks great, is there any blocker for this PR to get merged?

**Comment by hydrogen18** [comment]
@jay-dee7 This refactor is going to get merged into some other work I am doing to overhaul how we handle Kubernetes ingress objects and the assignment of hostnames to a users deployment on Akash

**Pull request closed without merging**